### PR TITLE
jmap_mail: load the cache correctly

### PR DESCRIPTION
**Pull request by brong**
Turns out we were calculating how much MORE to cache rather than what to cache up to, but the "loadn" API takes "up to here", not "this much more".  Doh.

### CODE REVIEW

**Comment by wolfsage** [comment]
I'll leave the code review to someone else - but for the commit message, can we include an explanation of what was wrong and how this fixes it?

Otherwise a year from now we'll just shake our heads at `jmap_mail: load the cache correctly` and wonder

Thanks!

**Comment by brong** [comment]
> I'll leave the code review to someone else - but for the commit message, can we include an explanation of what was wrong and how this fixes it?
> 
> Otherwise a year from now we'll just shake our heads at `jmap_mail: load the cache correctly` and wonder
> 
> Thanks!

Yep, done!  I've updated the commit message and also fixed the flipped logic of my last minute rewrite from the one I'd tested by hand on my testbed without good variable names.

File: imap/jmap_mail.c [hunk 1]
```diff
@@ -3858,24 +3858,29 @@ static void emailquery_cache_slice(struct emailquery *q,
...
-            if (lastpos > have_len) {
-                n = lastpos - have_len;
+            size_t have_len = q->collapse_threads ?
+                              qc->collapsed_len :
+                              qc->uncollapsed_len;
+            size_t need = q->super.position + q->super.limit;
+            // is the window already within the loaded section?
+            if (have_len < need) {
+                // fill just to the end of the window
+                emailquery_cache_loadn(q, qc, need);
```
Comment (rsto) [inline imap/jmap_mail.c:3878]: The `loadn` function is a misnomer I introduced, and now I can now how it is misleading. A better name might be `ensuren` or even `ensure`, like we do in the buf API.

**Review by rsto — APPROVED** [review]
Looks good to me, although we might want to rename `loadn`.

File: imap/jmap_mail.c (hunk 1 above)
Comment (brong) [reply imap/jmap_mail.c:3878]: I'll rename it!

**Review by rsto — APPROVED** [review]

**Pull request merged**